system/modules/system-partition.nix

```diff
@@ -53,5 +53,15 @@ in
       ln -s mnt/system/init init
       cd /
     '';
+
+    # Replace /dev/root by /dev/disk/by-label/system.x in /etc/fstab
+    boot.postBootCommands = with pkgs; ''
+      SYSTEM_DEVICE=$(readlink /dev/root)
+      mv /etc/fstab /etc/fstab.old
+      cp -L /etc/fstab.old /etc/fstab
+      chmod u+w /etc/fstab
+      ${gnused}/bin/sed -i "s|/dev/root|$SYSTEM_DEVICE|" /etc/fstab
+      chmod u-w /etc/fstab
+    '';
```

Comment:
Thanks for figuring out where things go wrong!

While this change resolves the symptoms, I am not sure it solves the underlying problem yet. I tried deleting the whole entry for `/dev/root` from `/etc/fstab` and that also leads to a functioning system.

I am also not quite sure the changelog entry from systemd is relevant. According to the docs, systemd will create dependencies on the mount units for entries from `fstab`:

https://github.com/systemd/systemd/blob/1e5d2d656420d0e755dbcf72aeba3c3aba54e956/man/systemd.mount.xml#L181-L188

I also don't yet understand what exactly fails, seeing that the `/dev/root` link is actually already established. It's all a bit mysterious.

Reply:
I need to come up from this rabbit hole ... :carrot: 

The whole phenomenon seems rather strange. There is a possibly related issue on the systemd project (https://github.com/systemd/systemd/issues/3446), with a panoply of different underlying problems and workarounds.

More workaround-heavy discussion here: https://unix.stackexchange.com/questions/556819/systemd-time-out-waiting-for-device-dev-mapper-vg

This rabbit hole might run deep and maybe your workaround is a good move in the end.